src/main/java/org/apache/gossip/manager/GossipCore.java

```diff
@@ -0,0 +1,243 @@
+      throw new RuntimeException(e);
+    }
+    int packet_length = json_bytes.length;
+    if (packet_length < GossipManager.MAX_PACKET_SIZE) {
+      byte[] buf = UdpUtil.createBuffer(packet_length, json_bytes);
+      try (DatagramSocket socket = new DatagramSocket()) {
+        InetAddress dest = InetAddress.getByName(uri.getHost());
+        DatagramPacket datagramPacket = new DatagramPacket(buf, buf.length, dest, uri.getPort());
+        socket.send(datagramPacket);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      } 
+    }
+  }
+  
+  public Response send(Base message, URI uri){
+    final Trackable t;
+    if (message instanceof Trackable){
+      t = (Trackable) message;
+      requests.put(t.getUuid(), message);
+    } else {
+      t = null;
+    }
+    sendInternal(message, uri);
+    if (t == null){
+      return null;
+    }
+    final CountDownLatch l = new CountDownLatch(1);
+    final AtomicReference<Response> r = new AtomicReference<>();
+    Thread thread = new Thread(){
+      public void run(){
+        while (true){
+          Base b = requests.get(t.getUuid() + "/" + t.getUriFrom());
+          if (b != null){
+            requests.remove(t.getUuid() + "/" + t.getUriFrom());
+          }
+          r.set((Response) b);
+          l.countDown();
+          try {
+            Thread.sleep(0, 1000);
+          } catch (InterruptedException e) {
+          }
+        }
+        
+      }
+    };
+    thread.start();
+    try {
+      l.await(10, TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    } finally {
+      requests.remove(t.getUuid() + "/" + t.getUriFrom());
+    }
+    return r.get();
+  }
+  
+  public void sendOneWay(Base message, URI u){
+    byte[] json_bytes;
+    try {
+      json_bytes = MAPPER.writeValueAsString(message).getBytes();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    int packet_length = json_bytes.length;
+    if (packet_length < GossipManager.MAX_PACKET_SIZE) {
+      byte[] buf = UdpUtil.createBuffer(packet_length, json_bytes);
+      try (DatagramSocket socket = new DatagramSocket()) {
+        InetAddress dest = InetAddress.getByName(u.getHost());
+        DatagramPacket datagramPacket = new DatagramPacket(buf, buf.length, dest, u.getPort());
+        socket.send(datagramPacket);
+      } catch (IOException ex) { }
+    }
+  }
+  
+
+  /**
+   * Merge remote list (received from peer), and our local member list. Simply, we must update the
+   * heartbeats that the remote list has with our list. Also, some additional logic is needed to
+   * make sure we have not timed out a member and then immediately received a list with that member.
+   * 
+   * @param gossipManager
+   * @param senderMember
+   * @param remoteList
+   * 
+   * COPIED FROM PASSIVE GOSSIP THREAD
+   */
+  protected void mergeLists(GossipManager gossipManager, RemoteGossipMember senderMember,
```

Comment:
Yep. I was having similar thoughts yesterday. One layering approach that came to mind was:

1. (lowest) transports (e.g. TCP, UDP).
2. Message marshaling.
3. State management.